### DISCUSSION TITLE
How To Get Table Name and Action When Using SQL Server Provider?

### DISCUSSION THREAD

**Discussion by chobo2** [thepirat000/Audit.NET#657] [Q&A]
Hi

I first started using the EF Provider but found out that it does not track low level changes, so I had to switch to the SqlServer provider but not sure how to set it up properly.

I had for EF

```
Audit.Core.Configuration.Setup()
.UseEntityFramework(config => config.AuditTypeMapper(t => typeof(EventLog)).AuditEntityAction<EventLog>((ev, entry, entity) =>
    {
    entity.TableName = entry.Table;
    entity.Action = entry.Action;
    }
).IgnoreMatchProperties(true));
```

Now I come up with this for SqlServer

```
Audit.Core.Configuration.Setup()
.UseSqlServer(config => config
.ConnectionString("string")
.TableName("EventLogs")
.IdColumName("Id")
.CustomColumn("Action", ev => ev.GetEntityFrameworkEvent() != null ? ev.GetEntityFrameworkEvent().Entiies.First().Action : ev.EventType)
.CustomColumn("TableName", ev => ev.GetEntityFrameworkEvent() != null ? ev.GetEntityFrameworkEvent().Entiies.First().Table : null)
)
```
I had to do some if statement checks and I found for the low level ones there is no Table name or action (Though EventType).

Is this the way I have to do it, or is there a better way of doing it?

**Comment by thepirat000** [comment]
I guess you mean the command interceptor events, right?.

The EF data provider cannot handle the command interceptor changes, but only the changes from SaveChanges calls. 

This is because the EF data provider relies on the EF change tracker which is not affected by commands.

One option could be to use a [conditional data provider wrapper](https://github.com/thepirat000/Audit.NET?tab=readme-ov-file#conditional-data-provider) to use EF data provider for the savechanges audits and sql data provider for the rest.

**Reply by chobo2** [reply]
Yes, as in the EF it says 

> Only the high-level audit events are processed by this data provider. Any other audit event, including the low-level events generated by the command interceptor, are ignored by the entity framework data provider.

I want to capture all information from select statement to inserts, deletes, updates and etc. 

What I want to do is have 2 columns in my "EventLogs" one for Tablename and one for  Action.

I was able to do that with EF easily, but not so easy with SQLServer provider.

Example
```
EventLogs Table
Id   Action TableName EventData
1   Insert   TableA    [json data of the event]
2.  Update TableB   [json data of the event]
3.  Select   TableA  [json data of the event]
4. Delete TableB  [json data of the event]
```
What I did puts "excuteReader" what I guess is fine if nothing else is available but nothing for table name as I can't find anything that would have it. 

So not sure if a conditional data provider wrapper would work? Maybe it would make it cleaner? Not sure if it would solve me from having to do .Entities.First() in a conditional check.

**Reply by thepirat000** [reply]
The command interceptor doesn't know about tables, it only knows the SQL command executed. If you need to extract the table(s) from the command you'll need to parse the command text

**Reply by thepirat000** [reply]
or maybe I'm wrong

https://stackoverflow.com/questions/73469589/how-to-get-table-name-from-idbcommandinterceptor-in-entityframework

Will take a deeper look when I've some time

**Reply by chobo2** [reply]
Cool, let me know if it can be done or not. Not a huge deal if it can't be done. 

As far I have my code to get the Action and a table name when it exists. Am I doing it the best way?

**Reply by thepirat000** [reply]
Your code will work only if you affect a single table on each SaveChanges() call

**Reply by thepirat000** [reply]
Looks like there is no way to retrieve the table name from an arbitrary DbCommand or SqlDbCommand for EF Core. 

Therefore, the only way to retrieve the table name is parsing the command text.

**Reply by chobo2** [reply]
Okay thanks for looking into it, not that big of a deal.  If that is the case you think it is better just to do a conditional wrapper or better to do your example or create a derived SQL data provider?

**Reply by thepirat000** [reply]
I believe the derived SQL data provider would give you more flexibility. Therefore, if it aligns with your requirements, I would recommend that option

**Comment by thepirat000** [comment]
For the high-level events, if you need to insert a record for each entry using the SqlDataProvider, I think a better way is to create a derived SQL data provider that inserts a different audit event for each EF entry. You could also create your own derived AuditEvent for this purpose.

For example:

```c#
public class SingleEntryAuditEvent : AuditEvent
{
    public EventEntry Entry { get; set; }
}
```

```c#
public class CustomSqlDataProvider : SqlDataProvider
{
    public CustomSqlDataProvider(Action<ISqlServerProviderConfigurator> config) : base(config) {}

    public override object InsertEvent(AuditEvent auditEvent)
    {
        foreach (var entry in auditEvent.GetEntityFrameworkEvent().Entries)
        {
            var entryEvent = CreateEventForEntry(auditEvent, entry);

            base.InsertEvent(entryEvent);
        }

        return null;
    }

    public override async Task<object> InsertEventAsync(AuditEvent auditEvent, CancellationToken cancellationToken = new CancellationToken())
    {
        foreach (var entry in auditEvent.GetEntityFrameworkEvent().Entries)
        {
            var entryEvent = CreateEventForEntry(auditEvent, entry);

            await base.InsertEventAsync(entryEvent, cancellationToken);
        }

        return null;
    }

    private static SingleEntryAuditEvent CreateEventForEntry(AuditEvent auditEvent, EventEntry entry)
    {
        var entryEvent = AuditEvent.FromJson<SingleEntryAuditEvent>(auditEvent.ToJson());
        entryEvent.Entry = entry;

        return entryEvent;
    }
}
```

So you can configure your data provider like this:

```c#
Audit.Core.Configuration.Setup()
    .Use(new CustomSqlDataProvider(config => config
        .ConnectionString("...")
        .TableName("EventLogs")
        .IdColumnName("Id")
        .CustomColumn("Action", ev => (ev as SingleEntryAuditEvent)?.Entry.Action ?? ev.EventType)
        .CustomColumn("TableName", ev => (ev as SingleEntryAuditEvent)?.Entry.Table)
    ));
```

**Reply by chobo2** [reply]
I not sure what I did wrong, I implemented your example, but now I don't see any "TableName" on any action (event on insert, update, delete) and for Action I just see DbContext instead of "Insert/Update/Delete".

Not sure if I missed something as I can't even debug this new class as it never gets triggered. I added CustomSqlDataProvider to mine.

**Reply by chobo2** [reply]
Is it because I got "JsonColumnName("Event")" that is causing the problem?

**Reply by chobo2** [reply]
This what I have (plus all your code now)

```
Audit.Core.Configuration.Setup()
    .Use(new CustomSqlDataProvider(config => config
        .ConnectionString("...")
        .TableName("EventLogs")
        .IdColumnName("Id")
      .CustomColumn("Date", ev => DateTime.Now)
        .CustomColumn("Action", ev => (ev as SingleEntryAuditEvent)?.Entry.Action ?? ev.EventType)
        .CustomColumn("TableName", ev => (ev as SingleEntryAuditEvent)?.Entry.Table)
    ));
```
I renamed my field to "entry" to see if that was the problem but now I get "entry" does not allow "null inserts"

I do have an EF Model 

```
public class AuditLog 
{
   public long Id {get; set;}
  public DateTime Date {get; set;}
 public required string Entry {get; set;}
// some other fields. Like TableName,(nullable) and Action (required)
}
```

**Reply by thepirat000** [reply]
Are you, by chance, also calling .UseEntityFramework()? You shouldn't. 

I don't understand the purpose of the AuditLog class. With the SQL data provider, you don't need an entity for the audit logs

**Reply by chobo2** [reply]
No I am not calling UseEntityFramework() as that will cause an override. Orignally I started by using that as I was planning to use EF provider but now that I need the low level events I had to change course. I right now do have both providers in my project but not using UseEntityFramework().

The reason why I have AuditLogs is because yeah I started with the EF provider. I kept it around right now as I want to access this "log" data in my application and I will be using EF to grab this data. 

No sure if that is causing a problem.

**Reply by thepirat000** [reply]
oh I see, that's not a problem.. The issue must be somewhere else.

Can you share a minimal project that reproduces the issue?

**Reply by chobo2** [reply]
Okay I think I forgot a section of your code. I only implemented the InsertEvent not the Async version of it.

Now I see my debugger. However the foreach gets a "null exception" when a low level event comes in (GetEntityFrameworkEvent()) seems to be null.

I can do a null check on auditEvent.GetEntityFrameworkEvent() but not sure if I should then do "auditEvent.EventType" in InsertEventAsync as I do see you have it in the CustomColumn but as it stands now that does not work.


```
 public override async Task<object> InsertEventAsync(AuditEvent auditEvent, CancellationToken cancellationToken = new CancellationToken())
    {
         if( auditEvent.GetEntityFrameworkEvent() == null)
         {
             // do something here?
         }
         
    
        foreach (var entry in auditEvent.GetEntityFrameworkEvent().Entries)
        {
            var entryEvent = CreateEventForEntry(auditEvent, entry);

            await base.InsertEventAsync(entryEvent, cancellationToken);
        }

        return null;
    }
```

**Reply by thepirat000** [reply]
Yeah, because the command interceptor generates a different kind of audit events, you can get the command information with `auditEvent.GetCommandEntityFrameworkEvent()` instead

**Reply by chobo2** [reply]
Do I need to loop that as well or use . result?

**Reply by chobo2** [reply]
Okay I played around wiht it and added

```
if(auditEvent.GetEntityFramework() == null )
{
   var e = AuditEvent.FromJson<SingleEntryAuditEvent>(auditEvent.ToJson());
  await base.InsertEventAsync(e)
}
```
that seems to now log the low level events, not sure if that is what you were thinking of doing?

I also have a few questions

1.  I don't need Audit.Entity.FrameworkCore package anymore right?
2.  If I want to opt out / in out of classes can I still do that or whas that only with the EF provider?

**Reply by thepirat000** [reply]
The Audit.EntityFramework.Core reference is still required. 

Note this package includes an audit event generator as well as a data provider. So, you will not be using the data provider anymore, but you are still auditing your DbContext.

You can use the optin/optout mechanism, but that will only apply to the high-level events.

**Reply by chobo2** [reply]
Okay, so opt out/in does not work for low level events?

Also, am I no longer able to use the Fluent Way to set up the Opt IN/OUT and must use the Data Annotation way? As I see the example is using the EF Configuration but if I would do that then it would override my provider correct?

**Reply by thepirat000** [reply]
You can use OptIn/OptOut with the fluent API, regardless of the data provider. 

```c#
Audit.EntityFramework.Configuration.Setup()
    .ForAnyContext(c => c.IncludeEntityObjects())
    .UseOptIn();
```

Please note that `Audit.EntityFramework.Configuration.Setup()` is for the EF high-level events configuration, distinct from the global configuration found under `Audit.Core.Configuration.Setup()`.

Also, note that the `AuditCommandInterceptor` will not take into account any configuration set via `Audit.EntityFramework.Configuration.Setup()`

**Reply by chobo2** [reply]
Okay, so doing what you put in the example plus what Audit.EntityFramework.Configuration.Setup() will not override each other. I thought as soon as you use either Audit.EntityFramework.Configuration.Setup() or Audit.Core.Configuration.Setup() one of them would be totally overridden.

**Reply by thepirat000** [reply]
Indeed, they are entirely independent configurations and do not overlap.
